### PR TITLE
Update AssetManager.php

### PR DESCRIPTION
**Pull request by nikitindv**
Add ability to define and override all assetsBundle options without necessarily enumeration of all used vendor bundles.

### CODE REVIEW

**Comment by samdark** [comment]
Could you elaborate what is it doing?

**Comment by nikitindv** [comment]
@samdark It allows me to override AssetBundle publish options for all included asset bundle, for example except by wildcard all *_spec.js files from BootstrapAsset.

**Comment by samdark** [comment]
Should be added to phpdoc / docs.

**Comment by nikitindv** [comment]
@samdark can i add pull request for documentation?
And if i do, in which language i shall do this?

**Comment by samdark** [comment]
Absolutely. Commit to the same `patch-1` branch. Commits will end up in this pull requests. Docs should be in English.

**Comment by nikitindv** [comment]
@samdark phpdoc updated.

File: framework/web/AssetManager.php [hunk 1]
```diff
@@ -260,6 +261,11 @@ protected function loadBundle($name, $config = [], $publish = true)
         if (!isset($config['class'])) {
             $config['class'] = $name;
         }
+        
+        if (isset($this->bundles['*'])) {
+            $config = array_merge($config, $this->bundles['*']);
```
Comment (samdark) [inline framework/web/AssetManager.php]: Should it be the other way around? i.e. what's specified in `$this->bundles['*']` is default values and what's passed to the method are more specific values.

Comment (nikitindv) [inline framework/web/AssetManager.php]: That argument requires same values as bundle specific value.
31 окт. 2015 г. 18:42 пользователь "Alexander Makarov" <
notifications@github.com> написал:

Comment (samdark) [inline framework/web/AssetManager.php]:
```
```

**Comment by ntemocentr** [comment]
Yes, there we merge bundle specific config with common config. So, if we define some bundle specific parameters, they will be applied as is, but if no - there will be used explicitly defined parameters from * key

**Comment by ntemocentr** [comment]
Offtopic: can we discuss my pull request in native language, and then translate for our foreign collegues, because my english is not good.

**Comment by samdark** [comment]
Yep. Catch me at https://gitter.im/yiisoft/yii2/rus

**Comment by klimov-paul** [comment]
I don't get it: why don't use a DI for this?

**Comment by samdark** [comment]
@klimov-paul example?

**Comment by klimov-paul** [comment]
`AssetBundle` instances are created via `Yii::createObject()`:

https://github.com/yiisoft/yii2/blob/master/framework/web/AssetManager.php#L264

Thus you should be able to configure thier properties via DI just the same way as we do this for widgets.
I can't see any obstacle for doing this at the first glance.

**Comment by samdark** [comment]
@ntemocentr ?

**Comment by klimov-paul** [comment]
In order to apply certain options for all Asset Bundles, you should be able to do something like following:

```php
$container->set('yii\web\AssetBundle', [
    'publishOptions' => [/* my options*/],
]);
```

**Comment by ntemocentr** [comment]
Imho it is not obvious for junior developers.
So many projects does not use assetic bundles and simpliy puts assets into webroot dir and uses plain links in templates.
DI is cool concept, but there must be simple way to override all default is arameters within any assetBundle, because assetsBundle is point extension.

**Comment by klimov-paul** [comment]
This does not sound like a convincing argument for me.
Problem of overring certain paramters for the several class of the same hierarchy or same functionality is solved long ago with the 'factory' OOP pattern, while "DI" pattern is an eveoled version of it.
I don't think the framework should introduce extra way to do such a thing, while DI alreay present.

**Comment by ntemocentr** [comment]
Thanks, using di a can introduce some syntax sugar within my projects)
I reached enlightenment, this pull request can be closed)

**Comment by ntemocentr** [comment]
@klimov-paul Hello, di doesn't allows me to override properties for inherited bundles, because of https://github.com/yiisoft/yii2/blob/master/framework/BaseYii.php#L341
\yii\BaseYii::createObject, does not look through class parents when applying properties.
So, i can't override BootstrapAsset props using 
$container->set('yii\web\AssetBundle', [
    'publishOptions' => [/* my options*/],
]);

**Comment by ntemocentr** [comment]
lets discuss this question in https://gitter.im/yiisoft/yii2/rus

**Comment by beowulfenator** [comment]
@klimov-paul, your argument in favor of DI pattern is valid. However, right now `AssetManager` has `bundles` property that allows to selectively override settings of certain asset bundles. This patch does not really introduce a new mechanism that has functionality similar to DI, it just extends existing functionality. After all, if we already allow to override settings of certain asset bundles in `AssetManager` config, what sense does it make to prohibit overriding them all?

**Comment by cebe** [comment]
This PR does not look like the right solution to the problem. Please open an issue to discuss possible solutions.

**Pull request closed without merging**